### DISCUSSION TITLE
cb-network v0.0.4 for Cafe Mocha release

### DISCUSSION THREAD

**Discussion by yunkon-kim** [cloud-barista/cb-larva#123] [Show and tell]
The cb-network's overall structure and control mechanism have been improved. The big changes included in this release are essential for cb-network to move towards supporting adaptiveness, fault-tolerance, lightweightness, and ease of use.

### Feature
- Major structural change of cb-network to the event-driven architecture
- Update configuration and control mechanisms of both cb-network server (controller) and agent for the followings:
  - Operate multiple Cloud Adaptive Networks (CLADNets)
  - Manage the overall CLADNet information on the distributed key-value store
  - Support concurrency
- Improve AdminWeb for the followings:
  - Support multiple Cloud Adaptive Networks (CLADNets)
  - Create the desired CLADNet CIDR block by the user
  - Check CLADNet connectivity and latency
- Add scripts to build and run agent for easy deployment
- Containerize cb-network server (controller)

#### Prerequisites
- A distributed key-value store clsuter is running
- The cb-network server (controller) is running

#### Tested with 
- [CB-Tumblebug v0.3.12](https://github.com/cloud-barista/cb-tumblebug/releases/tag/v0.3.12)
- [CB-Spider v0.3.19](https://github.com/cloud-barista/cb-spider/releases/tag/v0.3.19)

#### Test environment
NOTE - 1 VM is deployed in each region.

Cloud Service Provider (CSP) - Region pairs:
- **Amazon Web Services (AWS)** - Sydney, Australia
- **Alibaba Cloud** - Tokyo, Japan
- **Google Cloud Platform (GCP)** - Changhua County, Taiwan
- **Microsoft (MS) Azure** - California, United States

#### Note
- Use Golang v1.16
- Use Ubuntu 18.04 LTS (AWS, GCP, Alibaba Cloud) or 20.04 LTS (MS Azure)

#### Test scenario
- Create an MCIS (a logical group of VMs) by CB-Tumblebug API (Of course, CB-Spider is used.)
- Install and run the cb-network agent to the MCIS for configuring a CLADNet by CB-Tumblebug API
- Check if the CLADNet is successfully configured and running by Weave Scope

#### Test result
Source | Destination | Latency (ms)
-- | -- | --
192.168.10.3 (ali-ap-ne-1-0) | 192.168.10.4 (aws-ap-se-2-0) | 100.68
192.168.10.3 (ali-ap-ne-1-0) | 192.168.10.5 (az-wus-0) | 105.85
192.168.10.2 (gcp-as-east1-0) | 192.168.10.3 (ali-ap-ne-1-0) | 51.82
192.168.10.2 (gcp-as-east1-0) | 192.168.10.4 (aws-ap-se-2-0) | 146.80
192.168.10.2 (gcp-as-east1-0) | 192.168.10.5 (az-wus-0) | 149.12
192.168.10.4 (aws-ap-se-2-0) | 192.168.10.5 (az-wus-0) | 139.34

<hr /><em>This discussion was created from the release <a href='https://github.com/cloud-barista/cb-larva/releases/tag/v0.0.4'>cb-network v0.0.4 for Cafe Mocha release</a>.</em>

**Comment by seokho-son** [comment]
@hermitkim1 `v0.0.4` not `v0.4.0`. It is intentional, right?

**Reply by yunkon-kim** [reply]
@seokho-son 
Yes, it is 😃 
But actually, I'm not sure about the versioning so it could be changed until tomorrow.
